libservo: Starting removing the multi-document interface API

mrobinson opened the pull request:
For a while Servo has while had a somewhat unfinished multi-document
interface API that was meant to implement a kind of window manager
inside a single `RenderingContext`. This came with some focus handling
and a paint order implementation. This code was never really finished
and the `WebView` API is moving toward a design where compositing
different `WebView`s together is up to the embedder.

This is finally possible now that rendering to multiple
`RenderingContext`s is supported. In addition, the MDI API interferes
with focus handling when `WebView`s are distributed across windows.
Given these two points, this change starts to remove the MDI API.

- All `WebView`s can still be hidden / unhidden and start in an unhidden
  state.
- All `WebView`s always have system focus. System focus never properly
  interfaced with Servo's internal focus system anyway. A followup
  change will add proper system focus integration.
- There are still some leftovers from the MDI interface (such as
  `WebView`s having their own size). These will be cleaned up in a
  followup change. Only the changes necessary to get multi-window
  support working are included here.

Testing: This should not change observable behavior and tests are updated for the
new APIs.

mrobinson committed:
libservo: Starting removing the multi-document interface API

For a while Servo has while had a somewhat unfinished multi-document
interface API that was meant to implement a kind of window manager
inside a single `RenderingContext`. This came with some focus handling
and a paint order implementation. This code was never really finished
and the `WebView` API is moving toward a design where compositing
different `WebView`s together is up to the embedder.

This is finally possible now that rendering to multiple
`RenderingContext`s is supported. In addition, the MDI API interferes
with focus handling when `WebView`s are distributed across windows.
Given these two points, this change starts to remove the MDI API.

- All `WebView`s can still be hidden / unhidden and start in an unhidden
  state.
- All `WebView`s always have system focus. System focus never properly
  interfaced with Servo's internal focus system anyway. A followup
  change will add proper system focus integration.
- There are still some leftovers from the MDI interface (such as
  `WebView`s having their own size). These will be cleaned up in a
  followup change. Only the changes necessary to get multi-window
  support working are included here.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mukilan commented on ports/servoshell/egl/app.rs:
```diff
@@ -332,21 +332,19 @@ impl App {
-        if let Some(webview) = self.state.webview_by_id(id) {
-            webview.focus();
-        }
+    /// The active webview will be immediately valid via `active_or_newest_webview()`
```
Is the change in the semantics here intentional? I believe previous comment was highlighting the fact that focusing a webview is asynchronous.

mukilan commented on ports/servoshell/running_app_state.rs:
```diff
@@ -55,15 +55,20 @@ impl WebViewCollection {
         self.creation_order.retain(|&webview_id| webview_id != id);
-        if self.focused_webview_id == Some(id) {
-            self.focused_webview_id = None;
+        let removed_webivew = self.webviews.remove(&id);
```
```suggestion
        let removed_webview = self.webviews.remove(&id);
```

mukilan commented on components/servo/examples/winit_minimal.rs:
```diff
@@ -52,17 +52,6 @@ impl ::servo::WebViewDelegate for AppState {
         self.window.request_redraw();
     }
-
-    fn request_open_auxiliary_webview(&self, parent_webview: WebView) -> Option<WebView> {
```
Is this deletion deliberate?

mukilan commented:
I missed it during my testing for #40883, but on NixOS+Wayland, `Ctrl+Q` still triggers [this crash.](https://github.com/servo/servo/pull/40883#issuecomment-3584090609)

mrobinson commented on components/compositing/painter.rs:
```diff
@@ -792,9 +794,8 @@ impl Painter {
-        if self.webview_renderers.remove(webview_id).is_err() {
-            warn!("{webview_id}: Removing unknown webview");
+        if self.webview_renderers.remove(&webview_id).is_none() {
+            warn!("Tried removing uknown WebView: {webview_id:?}");
```
Adressed.

mrobinson commented on components/compositing/webview_renderer.rs:
```diff
@@ -102,6 +102,8 @@ pub(crate) struct WebViewRenderer {
     /// The HiDPI scale factor for the `WebView` associated with this renderer. This is controlled
     /// by the embedding layer.
     hidpi_scale_factor: Scale<f32, DeviceIndependentPixel, DevicePixel>,
+    /// Whether or not this [`WebViewRenderer`] is hidden or not.
```
Addressed.

mrobinson commented on components/constellation/constellation.rs:
```diff
@@ -4784,18 +4774,18 @@ where
+            .map(|webview| webview.focused_browsing_context_id);
+        focused_browsing_context_id.is_some_and(|focused_browsing_context_id| {
+            focused_browsing_context_id == change.browsing_context_id ||
+                self.fully_active_descendant_browsing_contexts_iter(change.browsing_context_id)
```
You're right. I've restored the original method name. I suspect that this code can be removed entirely. I don't know why loads should change focus.

mrobinson commented:
@mukilan Thanks for the review. I think I've addressed all of your review comments, apart from the one about the crash which I have opened #[40933](https://github.com/servo/servo/pull/40933)  for.
